Add RzBin plugin for Qualcomm peripheral images (`.mdt`).

Rot127 opened the pull request:
# DO NOT SQUASH

 <!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [x] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

I am sick of using the wonky Python scripts. And then load the whole pseudo-ELF binary, it doesn't resolve the nested ELF file(s) properly and just generally hurts my wish for order.

Also this allows for future extension, automatic decompression and such.

Besides this, it is just convenient to have.

Example, loading an `modem.mdt`. The symbols from the embedded ELF file (`modem.b32`) can be listed directly. No need to work with broken `ob`:

```
> rizin  modem.mdt 
WARNING: Failed to initialize section header.
 -- Print the contents of the current block with the 'p' command
[0x8b800000]> is
nth      paddr      vaddr bind   type   size lib name                                                  
-------------------------------------------------------------------------------------------------------
  1 0x000276d8 0x000266d8 GLOBAL FUNC     28     sys_sbrk_override
  2 0x0002a67c 0x0002967c GLOBAL FUNC     12     qurt_sem_destroy
  3 0x00038e68 0x00037e68 GLOBAL FUNC      4     smp2p_wait_multiple
  4 0x000281ec 0x000271ec GLOBAL FUNC     16     qurtos_reclaim_stack_pages
  5 0x00027978 0x00026978 GLOBAL FUNC     60     qurt_trampoline
  6 0x00026930 0x00025930 GLOBAL FUNC     24     qurt_mutex_destroy
  7 0x0002637c 0x0002537c GLOBAL FUNC     80     devcfg_blob_append
  8 0x00012a90 0x00011a90 GLOBAL FUNC     96     crc_16_step
  9 0x00036fd8 0x00035fd8 GLOBAL FUNC     24     qmi_idl_get_inherited_service_object
 10 0x00013354 0x00012354 GLOBAL FUNC     16     free
 11 0x0002748c 0x0002648c GLOBAL FUNC     72     sys_Mtxdst
...
```

Example sections of the embedded ELFs:

```
     paddr   size      vaddr  vsize align perm name                    type flags 
----------------------------------------------------------------------------------
0x00000000    0x0 0x0054ffff    0x0   0x0 ---- load-test.b03.0x0            
0x00001000 0x3ec0 0x00550000 0x3ec0   0x0 -rwx load-test.b03.start          
0x00005000   0x54 0x00554000   0x54   0x0 -r-x load-test.b03.init           
0x00006000 0x6088 0x00555000 0x6088   0x0 -r-x load-test.b03.text           
...
```

Example of `oml`:

```
 1 fd: 4 +0x00000000 0x89201c18 - 0x89201fff --- mmap.load-test.b01
 2 fd: 5 +0x00000000 0x89200000 - 0x89201c17 --- vmap.load-test.b01
 3 fd: 6 +0x00000000 0xfe030000 - 0xfe0aefff r-x mmap.load-test.b02
 4 fd: 7 +0x00000000 0xfe000000 * 0xfe02ffff r-x vmap.load-test.b02
 5 fd: 8 +0x00001000 0x00550000 - 0x00553ebf r-x vmap.load-test.b03.LOAD0
 6 fd: 8 +0x00005000 0x00554000 - 0x0055b763 r-x vmap.load-test.b03.LOAD1
 7 fd: 9 +0x00000000 0x0055d0a8 - 0x0055df03 rw- mmap.load-test.b03.LOAD2
 8 fd: 8 +0x0000d000 0x0055c000 - 0x0055d0a7 r-- vmap.load-test.b03.LOAD2
 9 fd: 10 +0x00000000 0xc80c94f4 - 0xc80c9567 r-- vmap.load-test.b04.reloc-targets
10 fd: 11 +0x00000388 0xc80c9388 - 0xc80c94e9 --- vmap.load-test.b04..strtab
11 fd: 12 +0x00000034 0xc80c9034 - 0xc80c90bb r-x vmap.load-test.b04..text
12 fd: 11 +0x0000022c 0xc80c922c - 0xc80c9387 --x vmap.load-test.b04..rela.text
13 fd: 11 +0x000000bc 0xc80c90bc - 0xc80c922b --- vmap.load-test.b04..symtab
14 fd: 11 +0x00000000 0xc80c9000 - 0xc80c9033 r-- vmap.load-test.b04.ehdr
```

Example of `iH`:

```
iH
==== MDT Segment 0 ====
     priv_p_flags: 0b00000111: layout | ELF
 -- ELF HEADER BEGIN -- 
0x00000000  MAGIC       7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00
0x00000010  Type        0x0002
0x00000012  Machine     0x00a4
0x00000014  Version     0x00000001
0x00000018  Entrypoint  0x8b800000
0x0000001c  PhOff       0x00000034
0x00000020  ShOff       0x00000000
0x00000024  Flags       0x0067
0x00000028  EhSize      52
0x0000002a  PhentSize   32
0x0000002c  PhNum       35
0x0000002e  ShentSize   40
0x00000030  ShNum       0
0x00000032  ShStrndx    0
 --- ELF HEADER END --- 

==== MDT Segment 1 ====
     priv_p_flags: 0b00000010: | MBN hash segment
 -- MBN AUTH HEADER BEGIN -- 
0x00 image_id:   kMbnImageDsp2 (0xc)
0x04 version:    0x6
0x08 paddr:      0x0
0x0c vaddr:      0x0
0x10 psize:      0x1f90
0x14 code_pa:    0x690
0x18 sign_va:    0x9b000488
0x1c sign_sz:    0x100
0x20 cert_va:    0x9b000488
0x24 cert_sz:    0x1800
 --- MBN AUTH HEADER END --- 

==== MDT Segment 2 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 3 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 4 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 5 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 6 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 7 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 8 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 9 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 10 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 11 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 12 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 13 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 14 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 15 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 16 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 17 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 18 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 19 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 20 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 21 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 22 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 23 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 24 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 25 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 26 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 27 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 28 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 29 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 30 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 31 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 32 ====
     priv_p_flags: 0b00001000: reloc | ELF
 -- ELF HEADER BEGIN -- 
0x00000000  MAGIC       7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00
0x00000010  Type        0x0003
0x00000012  Machine     0x00a4
0x00000014  Version     0x00000001
0x00000018  Entrypoint  0x00000000
0x0000001c  PhOff       0x00000034
0x00000020  ShOff       0x00072fc0
0x00000024  Flags       0x0067
0x00000028  EhSize      52
0x0000002a  PhentSize   32
0x0000002c  PhNum       4
0x0000002e  ShentSize   40
0x00000030  ShNum       25
0x00000032  ShStrndx    24
 --- ELF HEADER END --- 

==== MDT Segment 33 ====
     priv_p_flags: 0b00001000: reloc | Unidentified
==== MDT Segment 34 ====
     priv_p_flags: 0b00001000: reloc | Unidentified

```


Some questions to discuss:

This needs access to the ELF RzBin functions. Re-implementing all the ELF parsing doesn't make sense. But including everything a third or fourth time via `bin_elf.inc` (how it was done so far) is not ok either.
So I choose a middle ground here. I exposed versions of commonly used ELF parsing functions in `elf_parser`. They don't rely on `RzBinFile`, but take only the ELF object. This allows access for other `RzBin` plugins which rely on ELF files.

BUT, the design of the ELF module, only allows to be used either as ELF32 or ELF64. This is due to the ELF module being implemented twice with macros. This means every single module depending on it also is forced to copy itself (once for 32 and once for 64bit). Or am I missing something?

Unfortunately, due to this this bin loader only supports 32bit firmwares for now. Which is a shame, because these peripheral images are also used for AArch64 based sub-components.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Added

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

notxvilka commented:
>  This needs access to the ELF RzBin functions. Reimplementing everything doesn't make sense. But including everything a third or fourth time neither (the current approach is not ok IMHO). Are you finewith exposing some of the ELF parsers and getters as RZ_IPI. Generally this would allow access for other RzBin plugins which have nested ELF files.

You could see how it's done for PE and MDMP currently.

Rot127 commented:
> You could see how it's done for PE and MDMP currently.

PE has proper header files to include. So I would expose the elf functions as well like this.

Rot127 commented:
I open this for review now so you folks can give some early comment. Especially on the ELF thingy (see above). But no need to do a full review. It is mostly about the ELF thing and the overall mdt design.

notxvilka commented on librz/bin/bin.c:
```diff
@@ -1120,6 +1120,27 @@ RZ_API void rz_bin_virtual_file_free(RZ_NULLABLE RzBinVirtualFile *vfile) {
 	free(vfile);
 }
 
+RZ_API RZ_OWN RzBinVirtualFile *rz_bin_virtual_file_clone(RZ_BORROW RZ_NONNULL RzBinVirtualFile *vfile) {
```
Add Doxygen, brief explanation would be enough

notxvilka commented on librz/bin/bin.c:
```diff
@@ -1129,6 +1150,15 @@ RZ_API void rz_bin_map_free(RZ_NULLABLE RzBinMap *map) {
 	free(map);
 }
 
+RZ_API RZ_OWN RzBinMap *rz_bin_map_clone(RZ_NONNULL RzBinMap *map) {
```
Same

notxvilka commented on librz/bin/bin.c:
```diff
@@ -1129,6 +1150,15 @@ RZ_API void rz_bin_map_free(RZ_NULLABLE RzBinMap *map) {
 
+RZ_API RZ_OWN RzBinMap *rz_bin_map_clone(RZ_NONNULL RzBinMap *map) {
+	rz_return_val_if_fail(map, NULL);
+	RzBinMap *clone = RZ_NEW0(RzBinMap);
```
Check for NULL

notxvilka commented on librz/bin/format/elf/elf.h:
```diff
@@ -21,6 +21,8 @@
 #define RZ_BIN_ELF_PART_RELRO 1
 #define RZ_BIN_ELF_FULL_RELRO 2
 
+#define RZ_BIN_ELF_TINY_SIZE 45
```
Add a comment, it's not clear at all what does this mean

notxvilka commented on librz/bin/format/elf/elf_parser.c:
```diff
@@ -0,0 +1,205 @@
+
+#include "elf_parser.h"
+#include "bin/p/bin_elf.inc"
+
```
Add Doxygen `\file` explanation, parser name is quite generic

notxvilka commented on librz/bin/format/elf/elf_parser.c:
```diff
@@ -0,0 +1,205 @@
+	rz_bin_elf_foreach_relocs(obj, reloc) {
+		ut64 sym_addr = 0;
+		if (reloc->sym) {
+			RzBinElfSymbol *import = Elf_(rz_bin_elf_get_import)(obj, reloc->sym);
```
I suggest refactoring this piece of code to destroy the ladder

notxvilka commented on librz/bin/format/mbn/mbn.c:
```diff
@@ -0,0 +1,199 @@
+
+#include "mbn.h"
+#include <stddef.h>
+
```
Add `\file` to explain what `mbn` is. Also, missing copyright from you?

notxvilka commented on librz/bin/format/mbn/mbn.h:
```diff
@@ -0,0 +1,36 @@
+
+#define mbn_file_get_hdr(bf) ((SblHeader *)bf->o->bin_obj)
+
+typedef struct sbl_header {
```
If you have a reference - link to documentation, some article, etc would be nice to have it in the Doxygen description. Also please convert these comments  into Doxygen format too (maybe also add missing ones?)

notxvilka commented on librz/bin/format/mdt/mdt.c:
```diff
@@ -0,0 +1,653 @@
+	return part;
+}
+
+RZ_IPI void rz_bin_mdt_part_free(RzBinMdtPart *part) {
```
Not an API but I think it would be nice to add `RZ_OWN` to `_new()` and `RZ_NULLABLE` to `_free()`

notxvilka commented on librz/bin/format/mdt/mdt.c:
```diff
@@ -0,0 +1,653 @@
+	return elf_check_buffer_aux(b) == ELFCLASS32;
+}
+
+RZ_IPI bool rz_bin_mdt_check_buffer(RzBuffer *b) {
```
`rz_return_val_if_fail(b, false)`

notxvilka reviewed (changes requested):
Nice refactoring along the way! Please address my comments and it will be good to go.

notxvilka commented:
> BUT, the design of the ELF module, only allows to be used either as ELF32 or ELF64. This is due to the ELF module being implemented twice with macros. This means every single module depending on it also is forced to copy itself (once for 32 and once for 64bit). Or am I missing something?

Yes, I am afraid this is the case. ELF parser was refactored multiple times during its lifetime, gradually becoming more untangled and modular but there are still many warts left, including the need to extract relocation processing and parsing into a separate file at the very least. I think for now, you can do a copy and add support for 64-bit modules, we will deal with more refactoring later. It's already the case for the ELF after all.

Rot127 committed:
Expose ELF functions as API for other binary plugins.

It exposes commonly used ELF parsing functions in elf_parser.
They don't rely on RzBinFile but take only the ELF object.
This allows access for other RzBin plugins which rely on ELF files.

Rot127 committed:
Expose MBN plugin functions for other Bin plugins.

Rot127 committed:
Add the Qualcomm Peripheral Image loader (MDT files, 32bit only).

It loads .mdt firmware layout files, searches for the other parts in the same directory and loads them.
It resolves symbols, sections etc automatically, and maps them according to the layout
to the virtual address space.

- The iH command provides details about each firmware part as well.
- Memory maps and segment names are prefixed with the parts file name to make them distinct.
- If a firmware part is an ELF library it loads, and resolves its functions, segments etc.

Due to the design of the ELF binary plugin (code duplication with macros for 32/64bit versions),
this only supports 32bit firmware images for now.
This can relatively easily be fixed, by duplicating this plugin but including the 64bit ELF headers.
This requires more tests though.

Rot127 commented:
# DO NOT SQUASH
